### ISSUE TITLE
React 18 Typescript type error with formatMessage with values prop that takes a function returning a ReactElement

### ISSUE THREAD

**Issue by scottdickerson** [formatjs/formatjs#3856]
## Which package?
react-intl

## Describe the bug
A Typescript type error is thrown when passing a function that returns a ReactElement

Here's the example from the `react-intl` documentation that fails
```
function () {
  const messages = defineMessages({
    greeting: {
      id: 'app.greeting',
      defaultMessage: 'Hello, <bold>{name}</bold>!',
      description: 'Greeting to welcome the user to the app',
    },
  })

  return intl.formatMessage(messages.greeting, {
    name: 'Eric',
    bold: str => <b>{str}</b>,
  })
}
```

When you try this with `@types/react@18` you will get this type error:
```
No overload matches this call.
  Overload 1 of 2, '(descriptor: MessageDescriptor, values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>> | undefined, opts?: Options | undefined): string', gave the following error.
    Type '(str: string[]) => Element' is not assignable to type 'PrimitiveType | FormatXMLElementFn<string, string>'.
      Type '(str: string[]) => Element' is not assignable to type 'FormatXMLElementFn<string, string>'.
        Type 'Element' is not assignable to type 'string'.
  Overload 2 of 2, '(descriptor: MessageDescriptor, values?: Record<string, string | number | boolean | ReactElement<any, string | JSXElementConstructor<any>> | ... 5 more ... | undefined> | undefined, opts?: Options | undefined): ReactNode', gave the following error.
    Type '(str: string[]) => Element' is not assignable to type 'string | number | boolean | ReactElement<any, string | JSXElementConstructor<any>> | ReactFragment | ReactPortal | Date | FormatXMLElementFn<...> | null | undefined'.
      Type '(str: string[]) => Element' is not assignable to type 'FormatXMLElementFn<ReactNode, ReactNode>'.
        Types of parameters 'str' and 'parts' are incompatible.
          Type 'ReactNode[]' is not assignable to type 'string[]'.
            Type 'ReactNode' is not assignable to type 'string'.
              Type 'undefined' is not assignable to type 'string'.ts(2769)
```

### Reproducible Steps/Repo

1. Setup the example above.
2. Install  "@types/react": "18.0.21",

## Expected behavior
No typescript error occurs

## Desktop (please complete the following information):

- OS: OSX
- Browser N/A
- Version 
```
 "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-intl": "6.1.1",
```

**Comment by longlho** [comment]
cc @pyrocat101 as he's looking at this

**Comment by thomasboyt** [comment]
FYI - the fix in #3858 seems to require manually specifying a generic type allowing for a ReactNode return value, like so:

```ts
const formatted = intl.formatMessage<React.ReactNode>(message, {
  supportLink: (text) => <a href="/support">{text}</a>,
});
```

I'm a couple minor versions behind on react-intl so not sure if this is changed in newer versions. Works fine but a little bit hidden; wish https://formatjs.io/docs/react-intl/api#formatmessage had pointed it out (though, again, not sure if this is fixed on the latest version).
